services/app-api/handlers/templates/populateTemplatesTable.ts

```diff
@@ -0,0 +1,153 @@
+import { Handler } from "aws-lambda";
+import { AnyObject } from "yup/lib/types";
+import {
+  formTemplateTableName,
+  reportBuckets,
+} from "../../utils/constants/constants";
+import md5 from "md5";
+import s3Lib from "../../utils/s3/s3-lib";
+import dynamodbLib from "../../utils/dynamo/dynamodb-lib";
+import { FormTemplate } from "../../utils/types";
+import { S3 } from "aws-sdk";
+import * as path from "path";
+import { isFulfilled } from "../../utils/types/promises";
+import { logger } from "../../utils/logging";
+const REPORT_TYPES = ["MCPAR", "MLR"] as const;
+
+/**
+ * Retrieve template data from S3
+ *
+ * @param bucket s3 bucket
+ * @param key s3 key
+ * @returns s3 object body
+ */
+export async function getTemplate(bucket: string, key: string) {
+  return (await s3Lib.get({
+    Key: key,
+    Bucket: bucket,
+  })) as AnyObject;
+}
+
+/**
+ * Process a single form template, returning the form id and hash.
+ *
+ * @param bucket s3 bucket
+ * @param key s3 key
+ * @returns
+ */
+export async function processTemplate(bucket: string, key: string) {
+  const formTemplate = await getTemplate(bucket, key);
+  const hash = md5(JSON.stringify(formTemplate));
+  return {
+    id: path.basename(key).split(".")[0],
+    hash,
+  };
+}
+
+/**
+ * Create a distinct list of hashes.
+ *
+ * @param templates - list of { id, hash } objects
+ * @returns distinct list
+ */
+export function getDistinctHashesForTemplates(
+  templates: { id: string; hash: string }[]
+) {
+  return templates.filter(
+    (e, i) => templates.findIndex((a) => a["hash"] === e["hash"]) === i
+  );
+}
+
+/**
+ * Process a report type. This function retrieves all of the form templates in the database,
+ * sorts them by creation date, and removes duplicates.
+ * Each remaining form template becomes a new form version.
+ *
+ * @param reportType
+ */
+export async function processReport(reportType: typeof REPORT_TYPES[number]) {
+  const reportBucket = reportBuckets[reportType as keyof typeof reportBuckets];
+  const formTemplates = await s3Lib.list({ Bucket: reportBucket });
+
+  logger.info(`Found ${formTemplates.length} possible form template versions.`);
+
+  const sortedTemplates = formTemplates
+    .sort(
+      (a, b) =>
+        (b.LastModified?.getDate() ?? 0) - (a.LastModified?.getDate() ?? 0)
```

Comment:
Typo: `getDate()` returns the day of the month; we probably want `valueOf()`.

Thought: Subtracting the dates directly would also work (that is, `b.LastModified - a.LastModified`), if not for the special casing we would need to handle `undefined`. _Do_ we have any undefined LastModified dates? I can't imagine why we would... but it's probably not worth the effort to find out either.

Reply:
Oh thank you, good catch!

Our version of the AWS SDK types all of the object attributes (`LastModified`, `Key`, etc) as optional. Not sure why. Rather than mangling types (as I do later in the file), I just revert to 0 on the off chance the attribute is actually undefined.